Public/Src/Cache/MemoizationStore/Interfaces/Sessions/ContentHashList.cs

```diff
@@ -67,6 +68,36 @@ public class ContentHashList : IEquatable<ContentHashList>
             _payload = payload;
         }
 
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="ContentHashList" /> class.
+        /// </summary>
+        public ContentHashList(BinaryReader reader)
+        {
+            Contract.Requires(reader != null);
+            var numHashes = reader.ReadInt32();
+            if (numHashes < 0)
+            {
+                _contentHashes = null;
+            }
+            else
+            {
+                _contentHashes = new ContentHash[numHashes];

```

Comment:
>_contentHashes [](start = 16, length = 14)

I would strongly encourage to put serialization and deserialization logic close to each other.

It must be symetric, so it is easier to make sure this is the case if the methods are close to each otehr.